dali/operators/expressions/expression_impl_cpu.h

```diff
@@ -28,26 +27,18 @@
 
 namespace dali {
 
-// Assumes that it will get a workspace for given Backend,
-// and the backend will store the inputs and outputs at given Backend.
-template <ArithmeticOp op, typename Result,
-          typename Left, bool LeftIsTensor,
-          typename Right, bool RightIsTensor>
-class ExprImplBinCPU : public ExprImplBase, ExprImplParam<CPUBackend> {
+template <ArithmeticOp op, typename Result, typename Left, typename Right>
+class ExprImplCpuTT : public ExprImplBase {
  public:
-  ExprImplBinCPU() {}
-
-  void Execute(ArgumentWorkspace &workspace, const OpSpec &spec, ExprImplContext &ctx,
-               const std::vector<TileDesc> &tiles, TileRange range) override {
+  void Execute(ExprImplContext &ctx, const std::vector<ExtendedTileDesc> &tiles,
+               TileRange range) override {
     assert(range.begin + 1 == range.end &&
            "CPU Expression implementation can handle only one tile at a time");
-    auto &ws = dynamic_cast<workspace_t<CPUBackend> &>(workspace);
-    const auto &expr = dynamic_cast<const ExprFunc&>(*ctx.node);
     const auto &tile = tiles[range.begin];
-    auto left = ObtainInput<LeftIsTensor, Left>(expr, ws, spec, tile, 0);
-    auto right = ObtainInput<RightIsTensor, Right>(expr, ws, spec, tile, 1);
-    auto output = ObtainOutput<Result>(expr, ws, spec, tile);
-    Execute(output, left, right, tile.extent_size);
+    auto output = static_cast<Result *>(tile.output);
```

Comment:
Looks like passing a data through `void*`

Reply:
Yes. I see that indeed is what I thought it is.
Do we need to strip type?

Reply:
I would prefer not to, but that way we have a generic tile that works with all possible implementations.
We obtain the pointers for input & output from workspace (or from constant_storage). We can use the raw_data/raw_mutable_data, that are templateless and return a `void *` to the underlying buffer.
The ExprImpl knows what type it works with, so it knows to what type it should cast back. We got rid of separate Tile type for all the cases that way. I added a comment in `expression_tile.h` about it.